### DISCUSSION TITLE
Invalid polygons

### DISCUSSION THREAD

**Discussion by djes** [r-lidar/lidR#633] [Q&A]
Hello,

When using `crown_metrics()` or `delineate_crowns()` to create convex or concave polygons, I sometimes encounter the _'invalid polygon created. They likely correspond to trees with aligned points._' problem coming from the `metrics_crowns.R`  code, and processing halts. As I have tried everything to filter out problematic data and all parameters, I'm stuck to manual edition. 

Does someone have an idea? I've tried to create my own function but I'm not at ease with R.

I also wonder if it could be possible to use the `st_make_valid()` somewhere in the metrics_crowns.R code to, at least, not stop the processing ?

Regards,

J.B.

[crown bug example](https://drive.google.com/file/d/1y4w1EqWfNdMMXpW2gfN2QGyWjLzYrAdb/view?usp=share_link)

**Comment by Jean-Romain** [comment]
Please report a minimal reproducible example.

**Comment by djes** [comment]
Of course
```r
las <- readLAS("probleme.laz")
pol = crown_metrics(las, func = .stdmetrics, geom = "convex")
```
With the following file [probleme.laz](https://drive.google.com/file/d/1XuIPxqUbTlfLpek4yaVn5frS6EJA9_oB/view?usp=share_link)

**Reply by Jean-Romain** [reply]
I can't see any problem here. You have a indeed message. But it is only a message.

```r
valid <- sf::st_is_valid(pol)
pol[valid,]
```

**Reply by djes** [reply]
Sorry, I have not mentioned that the main problem is with a catalog, as the processing stops on the faulty tile and nothing is generated for the faulty tile. I can make a small example with a catalog if you want.

**Reply by Jean-Romain** [reply]
What is the error ? I mean, you only have a message here no error. Or maybe `sf` fails at writing the file ?

**Reply by djes** [reply]
Here is the error message
```
pol = crown_metrics(ctg, ~list(0, z_max = max(Z)), geom = "convex")

  |                                                                         
  |                                                                   |   0%
  |                                                                         
  |======================                                             |  33%Erreur : Evaluation error: IllegalArgumentException: Invalid number of points in LinearRing found 3 - must be 0 or >= 4.
```
I've rapidly created a small working example with 3 "small" faulty las files : https://drive.google.com/file/d/1y4w1EqWfNdMMXpW2gfN2QGyWjLzYrAdb/view?usp=share_link

**Reply by flottsam** [reply]
Did you ever find a solution? I'm receiving a similar error with warning message too: 

> scan() expected 'a real', got 'IllegalArgumentException:'
> 
> Error in any_list[[1]] : subscript out of bounds
> In addition: Warning message:
> 4 invalid polygons created. They likely correspond to trees with aligned points.
